### PR TITLE
Fix clang-tidy errors in headers

### PR DESCRIPTION
**Pull request by rdspring1**

### CODE REVIEW

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
Any thoughts on:
```
>>> Lint for csrc/utils.h:

  Error (CLANGTIDY) [clang-diagnostic-error]
    static_cast from 'nvfuser::PolymorphicBase *' to 'nvfuser::TensorView *',
    which are not related by inheritance, is not allowed

        183  |  template <class T>
        184  |  T* as() {
        185  |#ifdef NDEBUG
    >>> 186  |    auto downcast_ptr = static_cast<T*>(this);
        187  |#else
        188  |    auto downcast_ptr = dynamic_cast<T*>(this);
        189  |    TORCH_INTERNAL_ASSERT(downcast_ptr != nullptr);

  Error (CLANGTIDY) [clang-diagnostic-error]
    'const nvfuser::TensorView' is an incomplete type

        214  |  //!
        215  |  template <class T>
        216  |  bool isA() const {
    >>> 217  |    return dynamic_cast<const T*>(this) != nullptr;
        218  |  }
        219  |
        220  |  //! Check if the runtime type is strictly T. Returns false for classes
```
It seems to be a clang compile error, but doesn't appear when you run `lintrunner -a csrc/utils.h`

File: csrc/ir_internal_nodes.h [hunk 1]
```diff
@@ -414,13 +414,15 @@ class TORCH_CUDA_CU_API TernaryOp : public Expr {
...
 class TORCH_CUDA_CU_API RNGOp : public Expr {
-  size_t getOutputDims() const;
+  int64_t getOutputDims() const;
 
  public:
   struct Attributes {
-    RNGOpType rtype;
+    // default initialization for clang-tidy
+    // cppcoreguidelines-pro-type-member-init
+    RNGOpType rtype = RNGOpType::Uniform;
```
Comment (zasdfgbnm) [inline csrc/ir_internal_nodes.h:423]: Should we create a `RNGOpType::Undefined` for initialization?

**Comment by rdspring1** [comment]
`NOLINT` doesn't seem to work for clang-diagnostic-error.
https://stackoverflow.com/questions/50355930/ignore-clang-diagnostic-error-clang-tidy-caused-by-3rd-party-headers

**Comment by naoyam** [comment]
Which file triggers the error with `as`?

**Comment by rdspring1** [comment]
!build

**Comment by naoyam** [comment]
> I think you have to run `it_interface_node.h` and `utils.h` together for the error to appear
Let me try to fix this

**Comment by naoyam** [comment]
The error was because when just looking at, e.g., ir_internal_nodes.h, `TensorView` is not defined, so the compiler doesn't know if it's inherited from `PolymorphicBase`. Since `ir_interface_nodes.h` includes `ir_internal_nodes.h`, where `TensorView` is defined, we can't just include `ir_internal_nodes.h` from `ir_internal_nodes.h`. I pushed a workaround that basically moves all use of `as<TensorView>()` from the header to `ir_nodes.cpp`.

**Comment by naoyam** [comment]
> Does it make sense to change our coding preference just to make clang-tidy happy? I am OK with the change made here, but I am not sure if I will be happy to make similar changes in the future. For example, we might want to keep method definitions inside the header because it is inlined and has better performance. More importantly, if we define a template class, then we have no choice but to define all its methods in the header. In this case, I don't want to lose the flexibility of template class just because clang-tidy don't want to see it.

Does this really matter if functions are defined in header or not for inlining? I'd be strongly disappointed if any modern C++ compilers could not inline the functions I just moved to ir_nodes.cpp.

For templates, clang-tidy shouldn't complain if we only gave it a header file if there's no instantiation.

**Comment by rdspring1** [comment]
!build

**Comment by zasdfgbnm** [comment]
> I'd be strongly disappointed if any modern C++ compilers could not inline the functions I just moved to ir_nodes.cpp.

The C++ compiler definitely would want to inline, but they don't have enough information. For example, in `abc.cpp`, you included `ir_internal_nodes.h`, when compiling `abc.cpp`, the compiler has no information about the definition of these functions, so it just can not copy-paste the code into the callers when generating `abc.o`.

**Comment by zasdfgbnm** [comment]
> For templates, clang-tidy shouldn't complain if we only gave it a header file if there's no instantiation.

I feel like this is cheating, but 😃😃😃😃

**Comment by naoyam** [comment]
> The C++ compiler definitely would want to inline, but they don't have enough information. For example, in abc.cpp, you included ir_internal_nodes.h, when compiling abc.cpp, the compiler has no information about the definition of these functions, so it just can not copy-paste the code into the callers when generating abc.o.

Hahaha, my mental model is tainted with JIT dynamic compilations, and I forgot C++ only compiles one file at a time.

**Comment by naoyam** [comment]
Another thing to consider may be that the cpp files have been checked but nothing has been reported for the header files. Is it because our header files are recognized as system headers since we do `#include <....h>` rather than `#include "...h"`? If we instead do the latter, just running clang-tidy with the cpp files would be enough to detect errors in the header files too? 

When the header files are analyzed indirectly through the cpp files, we shouldn't need to do the changes I made.

**Comment by naoyam** [comment]
> > For templates, clang-tidy shouldn't complain if we only gave it a header file if there's no instantiation.
> 
> I feel like this is cheating, but 😃😃😃😃

It's not really cheating since we still run clang-tidy with cpp files where templates are instantiated. So, all the uses of templates should get checked.

**Comment by zasdfgbnm** [comment]
> Another thing to consider may be that the cpp files have been checked but nothing has been reported for the header files. Is it because our header files are recognized as system headers since we do `#include <....h>` rather than `#include "...h"`? If we instead do the latter, just running clang-tidy with the cpp files would be enough to detect errors in the header files too?
> 
> When the header files are analyzed indirectly through the cpp files, we shouldn't need to do the changes I made.

I don't think this will work. Actually, `scheduler/matmul_utils.cpp` is already using `#include ""` to include that header (which I think needs to be fixed for consistent coding style BTW), and I just ran `lintrunner`, it doesn't help.

**Comment by naoyam** [comment]
> > Another thing to consider may be that the cpp files have been checked but nothing has been reported for the header files. Is it because our header files are recognized as system headers since we do `#include <....h>` rather than `#include "...h"`? If we instead do the latter, just running clang-tidy with the cpp files would be enough to detect errors in the header files too?
> > When the header files are analyzed indirectly through the cpp files, we shouldn't need to do the changes I made.
> 
> I don't think this will work. Actually, `scheduler/matmul_utils.cpp` is already using `#include ""` to include that header (which I think needs to be fixed for consistent coding style BTW), and I just ran `lintrunner`, it doesn't help.

Haven't looked into it deeply, but looks like it first reads `ir_interface_nodes.h`, which reads `ir_internal_nodes.h` as a system header, so that may be the reason you didn't see any error with `ir_internal_nodes.h`

**Comment by zasdfgbnm** [comment]
> > > Another thing to consider may be that the cpp files have been checked but nothing has been reported for the header files. Is it because our header files are recognized as system headers since we do `#include <....h>` rather than `#include "...h"`? If we instead do the latter, just running clang-tidy with the cpp files would be enough to detect errors in the header files too?
> > > When the header files are analyzed indirectly through the cpp files, we shouldn't need to do the changes I made.
> > 
> > 
> > I don't think this will work. Actually, `scheduler/matmul_utils.cpp` is already using `#include ""` to include that header (which I think needs to be fixed for consistent coding style BTW), and I just ran `lintrunner`, it doesn't help.
> 
> Haven't looked into it deeply, but looks like it first reads `ir_interface_nodes.h`, which reads `ir_internal_nodes.h` as a system header, so that may be the reason you didn't see any error with `ir_internal_nodes.h`

I don't think this is the case, just swapped the order, still not working :(

**Comment by naoyam** [comment]
Maybe need this? `-header-filter='csrc/*'`

**Comment by zasdfgbnm** [comment]
> Maybe need this? `-header-filter='csrc/*'`

Hmm, I don't think lintrunner supports taking additional arguments for clang-tidy.

**Comment by naoyam** [comment]
That's unfortunate as I see the errors fixed in this PR indeed showing up on `main` with this command:

```
clang-tidy-13 -header-filter='csrc/*' csrc/ir_nodes.cpp 
```

**Comment by rdspring1** [comment]
You should be able to add extra arguments here.
https://github.com/NVIDIA/Fuser/blob/main/tools/linter/adapters/clangtidy_linter.py#LL148C14-L148C20

**Comment by zasdfgbnm** [comment]
I think we would need both `-header-filter='csrc/*'` and the change in https://github.com/NVIDIA/Fuser/pull/331? If we were always checking all cpp files, then https://github.com/NVIDIA/Fuser/pull/331 shouldn't matter, but unfortunately we are only checking files change by PR. And if a PR only changes a header file, then that header file should still be checked. But unfortunately, as long as we check headers standalone, we will have the issue of unable to put function definition in a header.

**Comment by naoyam** [comment]
Hmm, you're right.

If this `ir_internal_nodes.h` and its uses of `as` and `isA` are the only problem, we could also do a little dirty hack, like:

```
diff --git a/csrc/utils.h b/csrc/utils.h
index eec0038..0e6f10c 100644
--- a/csrc/utils.h
+++ b/csrc/utils.h
@@ -182,6 +182,9 @@ class PolymorphicBase {
   // (checked in DEBUG builds)
   template <class T>
   T* as() {
+#ifdef CLANG_TIDY
+      return nullptr;
+#else
 #ifdef NDEBUG
     auto downcast_ptr = static_cast<T*>(this);
 #else
@@ -189,10 +192,14 @@ class PolymorphicBase {
     TORCH_INTERNAL_ASSERT(downcast_ptr != nullptr);
 #endif
     return downcast_ptr;
+#endif
   }

   template <class T>
   const T* as() const {
+#ifdef CLANG_TIDY
+      return nullptr;
+#else
 #ifdef NDEBUG
     auto downcast_ptr = static_cast<const T*>(this);
 #else
@@ -200,6 +207,7 @@ class PolymorphicBase {
     TORCH_INTERNAL_ASSERT(downcast_ptr != nullptr);
 #endif
     return downcast_ptr;
+#endif
   }

   //! Check if the runtime type is T (or derived from T)
@@ -214,7 +222,11 @@ class PolymorphicBase {
   //!
   template <class T>
   bool isA() const {
+#ifdef CLANG_TIDY
+      return true;
+#else
     return dynamic_cast<const T*>(this) != nullptr;
+#endif
   }
```

Then run clang-tidy as `clang-tidy-13 --extra-arg=-DCLANG_TIDY csrc/ir_internal_nodes.h`. Maybe this is better than putting some small functions in ir_nodes.cpp?

**Comment by zasdfgbnm** [comment]
Yeah, I prefer your new approach, although it is also very ugly... For `as`, could we use `reinterpret_cast` to replace `static_cast`? Not sure if clang-tidy will allow us to do so...

**Comment by naoyam** [comment]
> For as, could we use reinterpret_cast to replace static_cast? Not sure if clang-tidy will allow us to do so...

Yes, clang-tidy seems happy with `reinterpret_cast`, but I don't think we want to do so for the normal case as it would mean no type check is done.

**Comment by naoyam** [comment]
Rethinking about this, I wonder how the use of `as<TensorView>()` in ir_internal_nodes.h could work since when reading the header file, there's only a forward declaration of `TensorView`, so how can the compiler reason about `static_cast<TensorView*>()`?

**Comment by zasdfgbnm** [comment]
> Rethinking about this, I wonder how the use of `as<TensorView>()` in ir_internal_nodes.h could work since when reading the header file, there's only a forward declaration of `TensorView`, so how can the compiler reason about `static_cast<TensorView*>()`?

I think `static_cast` is the same as `reinterpret_cast` for pointers. It's just clang-tidy is unreasonably happy with one but unhappy with the other.

**Comment by rdspring1** [comment]
My take was that we have a circular dependency between `ir_interface_nodes.h` and `ir_internal_nodes.h`.
- `ir_internal_nodes.h` depends on `TensorView`
- `ir_interface_nodes.h` depends on `IterDomain` and `TensorDomain`

My solution is to move `IterDomain` and `TensorDomain` to a separate header file higher in the hierarchy. We don't have to move methods to the cpp file in this case.

Here is the proposed header hierarchy:
- `utils.h` - `PolymorphicBase` and `NonCopyable`
- `ir_base_nodes.h` - `Statement, Expr, Val`
-  `ir_internal_base_nodes.h` - `IterDomain and TensorDomain` 
--- **TODO: Move `IterDomain and TensorDomain` to separate header file**
- `ir_interface_nodes.h` - `TensorView`
- `ir_internal_nodes.h`

**Comment by rdspring1** [comment]
!build

**Comment by naoyam** [comment]
Sounds good to me. Thanks for the refactoring.

**Comment by zasdfgbnm** [comment]
@rdspring1 I like your approach.

**Comment by naoyam** [comment]
@rdspring1 I see you already removed `IterDomain` from ir_internal_nodes.h, but I don't see `ir_internal_base_nodes.h`. Have you added the file?

**Comment by rdspring1** [comment]
`IterDomain` is in `ir_base_nodes.h` right now. I'm moving them to `ir_internal_base_nodes.h` now.

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
@naoyam I moved the files.

**Review by zasdfgbnm — APPROVED** [review]
LGTM, thanks for these fixes!

**Review by naoyam — APPROVED** [review]
Thanks!

**Pull request merged**